server/src/test/java/org/elasticsearch/index/translog/TranslogTests.java

```diff
@@ -1507,6 +1511,302 @@ public void testSnapshotFromStreamInput() throws IOException {
+            translog.add(operation);
+        }
+        try {
+            translog.trimOperations(pt, 0);
+            fail();
+        } catch (AssertionError e) {
+            assertThat(e.getMessage(),
+                is("current should not have any operations with seq#:primaryTerm [1:" + (pt - 1) + "] > 0:" + pt));
+        }
+
+        primaryTerm.incrementAndGet();
+        translog.rollGeneration();
+
+        // add a single operation to current with seq# > trimmed seq# but higher primary term
+        Translog.Index operation = new Translog.Index("test", "" + 1, 1L, primaryTerm.get(),
+            randomAlphaOfLengthBetween(1, 50).getBytes("UTF-8"));
+        translog.add(operation);
+
+        // it is possible to trim after generation rollover
+        translog.trimOperations(pt, 0);
+    }
+
+    public void testSnapshotTrimmedOperations() throws Exception {
+        List<Translog.Index> allOperations = new ArrayList<>();
+        List<Translog.Index> effectiveOperations = new ArrayList<>();
+
+        long translogOperations = 0;
+        final AtomicBoolean wasRollover = new AtomicBoolean(false);
+
+        for(int attempt = 0, maxAttempts = randomIntBetween(3, 10); attempt < maxAttempts; attempt++) {
+            int extraDocs = randomIntBetween(10, 15);
+
+            List<Translog.Index> newOperations = new ArrayList<>();
+
+            List<Long> ops = LongStream.range(0, translogOperations + extraDocs)
+                .boxed().collect(Collectors.toList());
+            Randomness.shuffle(ops);
+
+            for (final long op : ops) {
+                // use ongoing primaryTerms - or the same as it was
+                final long pt = this.primaryTerm.get();
+
+                String source = randomAlphaOfLengthBetween(1, 50);
+
+                // if there was no roll generation - have to use exactly the same source
+                // for same seq# regardless primaryTerm
+                if (!wasRollover.get()) {
+                    for (Translog.Index allOp : allOperations) {
+                        if (allOp.seqNo() == op) {
+                            // use the latest source of op with the same seq# - therefore no break
+                            source = allOp.source().utf8ToString();
+                        }
+                    }
+                }
+
+                Translog.Index operation = new Translog.Index("test", "" + op, op, pt, source.getBytes("UTF-8"));
+                translog.add(operation);
+
+                allOperations.add(operation);
+                newOperations.add(operation);
+            }
+
+            wasRollover.set(randomBoolean());
+            if (wasRollover.get()) {
+                primaryTerm.incrementAndGet();
+                translog.rollGeneration();
+            }
+
+            int trimmedDocs = randomIntBetween(4, 8);
+            long maxTrimmedSeqNo = wasRollover.get() ? translogOperations + extraDocs - trimmedDocs : translogOperations + extraDocs + 1;
```

Comment:
why do we need to look at the rollover to determine this parameter? it feels weird to rely on internal semantics for an external operations.